File: strategy/signal/apo_flip.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    StaticParameter,
)
from strategy.signal.base import Signal, SignalType


@dataclass(frozen=True)
class APOFlipSignal(Signal):
    type: SignalType = SignalType.ApoFlip
    short_period: Parameter = StaticParameter(10.0)
    long_period: Parameter = StaticParameter(20.0)

File: ta_lib/strategies/signal/src/apo_flip.rs
use base::{OHLCVSeries, Signal};
use core::{Cross, Series};
use momentum::apo;

const APO_ZERO: f32 = 0.0;

pub struct APOFlipSignal {
    short_period: usize,
    long_period: usize,
}

impl APOFlipSignal {
    pub fn new(short_period: f32, long_period: f32) -> Self {
        Self {
            short_period: short_period as usize,
            long_period: long_period as usize,
        }
    }
}

impl Signal for APOFlipSignal {
    fn lookback(&self) -> usize {
        std::cmp::max(self.short_period, self.long_period)
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let ao = apo(&data.close, self.short_period, self.long_period);

        (ao.cross_over(APO_ZERO), ao.cross_under(APO_ZERO))
    }
}
